core/state/trie_prefetcher.go

```diff
@@ -246,139 +309,307 @@ type subfetcher struct {
-	// No matter how the loop stops, signal anyone waiting that it's terminated
-	defer close(sf.term)
+func (sf *subfetcher) skips() int {
+	if sf.to == nil {
+		// Unable to open trie
+		return 0
+	}
+	return int(sf.to.skips.Load())
+}
 
+func (sf *subfetcher) copies() int {
+	if sf.to == nil {
+		// Unable to open trie
+		return 0
+	}
+	return sf.to.copies
+}
+
+// trieOrchestrator is not thread-safe.
+type trieOrchestrator struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	sf *subfetcher
+
+	// base is an unmodified Trie we keep for
+	// creating copies for each worker goroutine.
+	//
+	// We care more about quick copies than good copies
+	// because most (if not all) of the nodes that will be populated
+	// in the copy will come from the underlying triedb cache. Ones
+	// that don't come from this cache probably had to be fetched
+	// from disk anyways.
+	base     Trie
+	baseLock sync.Mutex
+
+	outstandingRequests sync.WaitGroup
+	tasksAllowed        bool
+	skips               atomic.Int32 // Number of tasks skipped
+	pendingTasks        [][]byte
+	pendingTasksLock    sync.Mutex
+	wake                chan struct{}
+	loopTerm            chan struct{}
+
+	copies      int
+	copyChan    chan Trie
+	copySpawner chan struct{}
+}
+
+func newTrieOrchestrator(sf *subfetcher) *trieOrchestrator {
 	// Start by opening the trie and stop processing if it fails
+	var (
+		base Trie
+		err  error
+	)
 	if sf.owner == (common.Hash{}) {
-		trie, err := sf.db.OpenTrie(sf.root)
+		base, err = sf.db.OpenTrie(sf.root)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return
+			return nil
 		}
-		sf.trie = trie
 	} else {
-		trie, err := sf.db.OpenStorageTrie(sf.state, sf.owner, sf.root)
+		base, err = sf.db.OpenStorageTrie(sf.state, sf.owner, sf.root)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return
+			return nil
 		}
-		sf.trie = trie
 	}
 
-	// Trie opened successfully, keep prefetching items
+	// Create cancelable context
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Instantiate trieOrchestrator
+	to := &trieOrchestrator{
+		ctx:       ctx,
+		ctxCancel: cancel,
+
+		sf:   sf,
+		base: base,
+
+		tasksAllowed: true,
+		wake:         make(chan struct{}, 1),
+		loopTerm:     make(chan struct{}),
+
+		copyChan:    make(chan Trie, sf.p.maxConcurrency),
+		copySpawner: make(chan struct{}, sf.p.maxConcurrency),
+	}
+
+	// Create initial trie copy
+	to.copies++
+	to.copySpawner <- struct{}{}
+	to.copyChan <- to.copyBase()
+	return to
+}
+
+func (to *trieOrchestrator) copyBase() Trie {
+	to.baseLock.Lock()
+	defer to.baseLock.Unlock()
+
+	return to.sf.db.CopyTrie(to.base)
+}
+
+func (to *trieOrchestrator) enqueueTasks(tasks [][]byte) {
+	if len(tasks) == 0 {
+		return
+	}
+
+	// Add tasks to [pendingTasks]
+	to.pendingTasksLock.Lock()
+	if !to.tasksAllowed {
+		to.skips.Add(int32(len(tasks)))
```

Comment:
can probably simplify code by putting a single lock around `skips`, `pendingTasks`, `outstandingRequests` instead of using a combination of locks/atomics.